pennylane/ops/op_math/prod.py

```diff
@@ -276,20 +286,31 @@ def diagonalizing_gates(self):
         Returns:
             list[.Operator] or None: a list of operators
         """
-
-        eigen_vectors = self.eigendecomposition["eigvec"]
-        return [qml.QubitUnitary(eigen_vectors.conj().T, wires=self.wires)]
+        if self.overlapping_wires:
+            eigen_vectors = self.eigendecomposition["eigvec"]
+            return [qml.QubitUnitary(eigen_vectors.conj().T, wires=self.wires)]
+        diag_gates = []
+        for factor in self.factors:
+            diag_gates.extend(factor.diagonalizing_gates())
+        return [qml.adjoint(gate) for gate in diag_gates]
```

Comment:
Let *O<sub>A</sub> = U<sub>A</sub>Λ<sub>A</sub>U†<sub>A</sub>* and *O<sub>B</sub> = U<sub>B</sub>Λ<sub>B</sub>U†<sub>B</sub>* be 2 operators that act on different Hilbert spaces.

Then:
*O<sub>A</sub>O<sub>B</sub> = U<sub>A</sub>Λ<sub>A</sub>U†<sub>A</sub>U<sub>B</sub>Λ<sub>B</sub>U†<sub>B</sub> = U<sub>A</sub>U<sub>B</sub>Λ<sub>A</sub>Λ<sub>B</sub>U†<sub>A</sub>U†<sub>B</sub>*

U<sub>A</sub>U<sub>B</sub> is the product's diagonalising gate. The eigenvalues are the product of the expanded eigenvalues of each factor.